File: public/code/chartpatterns__rectangle.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def pivots(c, lb=5):
    sh = swingHighs(c, lb)
    sl = swingLows(c, lb)
    highs = []
    lows = []
    for i in range(0, len(c)):
        if sh[i] != None:
            highs.append({'idx': i, 'price': sh[i]})
        if sl[i] != None:
            lows.append({'idx': i, 'price': sl[i]})
    return {'highs': highs, 'lows': lows}
def rectangle(c):
    if len(c) < 50:
        return make_signal(reason="Insufficient data")
    _d = pivots(c, 4)
    highs = _d['highs']
    lows = _d['lows']
    if len(highs) < 2  or  len(lows) < 2:
        return make_signal(reason="Not enough pivots")
    recentH = highs[-2:]
    recentL = lows[-2:]
    a = atr(c, 14)
    i = len(c) - 1
    cur = c[i]
    resistance = sum(h.price for h in recentH) / len(recentH)
    support = sum(l.price for l in recentL) / len(recentL)
    # Tepeler ve dipler düz (rectangle)
    flatRes = abs(recentH[0].price - recentH[1].price) / resistance < 0.015
    flatSup = abs(recentL[0].price - recentL[1].price) / support < 0.015
    if not flatRes  or  not flatSup:
        return make_signal(reason="No rectangle")
    height = resistance - support
    if cur['close'] > resistance  and  c[i - 1]['close'] <= resistance:
        return make_signal(signal="long", entry=cur['close'], stop_loss=resistance - height * 0.5, take_profit=[cur['close'] + height * 0.5, cur['close'] + height, cur['close'] + height * 1.5], confidence=0.71, reason="Rectangle breakout up")
    if cur['close'] < support  and  c[i - 1]['close'] >= support:
        return make_signal(signal="short", entry=cur['close'], stop_loss=support + height * 0.5, take_profit=[cur['close'] - height * 0.5, cur['close'] - height, cur['close'] - height * 1.5], confidence=0.71, reason="Rectangle breakdown down")
    return make_signal(reason="Price inside rectangle")

File: public/code/chartpatterns__rectangle.ts
import { Candle, Signal, makeSignal, atr, sma, swingHighs, swingLows } from "../indicators";

function pivots(c: Candle[], lb = 5) {
  const sh = swingHighs(c, lb), sl = swingLows(c, lb);
  const highs: { idx: number; price: number }[] = [], lows: { idx: number; price: number }[] = [];
  for (let i = 0; i < c.length; i++) {
    if (sh[i] !== null) highs.push({ idx: i, price: sh[i] as number });
    if (sl[i] !== null) lows.push({ idx: i, price: sl[i] as number });
  }
  return { highs, lows };
}

export function rectangle(c: Candle[]): Signal {
  if (c.length < 50) return makeSignal({ reason: "Insufficient data" });
  const { highs, lows } = pivots(c, 4);
  if (highs.length < 2 || lows.length < 2) return makeSignal({ reason: "Not enough pivots" });
  const recentH = highs.slice(-2), recentL = lows.slice(-2);
  const a = atr(c, 14), i = c.length - 1, cur = c[i];
  const resistance = recentH.reduce((s, h) => s + h.price, 0) / recentH.length;
  const support = recentL.reduce((s, l) => s + l.price, 0) / recentL.length;
  // Tepeler ve dipler düz (rectangle)
  const flatRes = Math.abs(recentH[0].price - recentH[1].price) / resistance < 0.015;
  const flatSup = Math.abs(recentL[0].price - recentL[1].price) / support < 0.015;
  if (!flatRes || !flatSup) return makeSignal({ reason: "No rectangle" });
  const height = resistance - support;
  if (cur.close > resistance && c[i - 1].close <= resistance) {
    return makeSignal({ signal: "long", entry: cur.close, stop_loss: resistance - height * 0.5, take_profit: [cur.close + height * 0.5, cur.close + height, cur.close + height * 1.5], confidence: 0.71, reason: "Rectangle breakout up" });
  }
  if (cur.close < support && c[i - 1].close >= support) {
    return makeSignal({ signal: "short", entry: cur.close, stop_loss: support + height * 0.5, take_profit: [cur.close - height * 0.5, cur.close - height, cur.close - height * 1.5], confidence: 0.71, reason: "Rectangle breakdown down" });
  }
  return makeSignal({ reason: "Price inside rectangle" });
}
